### PR TITLE
Fix a bunch of parser/formatter bugs found in fuzzing

### PR DESCRIPTION
**Pull request by joshuawarner32**
Notably:
* Unified how parens are formatted between (1) when we have a ParensAround, and (2) when we've decided an Apply needs to have parens
* Made unary minus require the be indented to the same level as any other expression continuation. (it used to accidentally have rules meant for binary operators applied)
* Don't apply extra indent to the backpassing continuation in the case that the call does itself require indentation
* Make `try@foo` correctly parse as `try @foo`, so that formatting doesn't change the tree when it adds that space
* Detect more cases where we need to outdent trailing e.g. {} blocks in applies
* Approximately a bagillion other things, 90% of which I added tests for, and none of which affected the formatting of examples or builtins

### CODE REVIEW

File: crates/compiler/fmt/src/annotation.rs [hunk 1]
```diff
@@ -704,3 +824,265 @@ pub fn except_last<T>(items: &[T]) -> impl Iterator<Item = &T> {
...
+        TypeAnnotation::SpaceAfter(expr, spaces) => {
+            let mut inner = ann_lift_to_node(parens, arena, expr);
+            inner.after = merge_spaces_conservative(arena, inner.after, spaces);
+            inner
+        }
+        TypeAnnotation::Function(_args, _purity, _res) => {
+            let inner = Spaces {
+                before: &[], // TODO: grab the spaces from the args
+                item: Node::TypeAnnotation(*ann),
+                after: &[], // TODO: grab the spaces from the res
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/annotation.rs:1027]: TODO's -- for follow up, or missed?

File: crates/compiler/fmt/src/def.rs [hunk 2]
```diff
@@ -51,6 +69,440 @@ impl<'a> Formattable for Defs<'a> {
...
+                before: &[],
+                item: ValueDef::Expect {
+                    condition: arena.alloc(Loc::at(condition.region, condition_lifted.item)),
+                    preceding_comment,
+                },
+                after: condition_lifted.after,
+            }
+        }
+        ValueDef::ModuleImport(module_import) => {
+            // TODO!
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/def.rs:267]: TODO

File: crates/compiler/fmt/src/def.rs [hunk 3]
```diff
@@ -51,6 +69,440 @@ impl<'a> Formattable for Defs<'a> {
...
+            Spaces {
+                before: &[],
+                item: ValueDef::ModuleImport(module_import),
+                after: &[],
+            }
+        }
+        ValueDef::IngestedFileImport(ingested_file_import) => {
+            // TODO!
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/def.rs:275]: TODO

File: crates/compiler/fmt/src/def.rs [hunk 4]
```diff
@@ -51,6 +69,440 @@ impl<'a> Formattable for Defs<'a> {
...
+            preceding_comment,
+        } => SpacesBefore {
+            before: &[],
+            item: ValueDef::Expect {
+                condition,
+                preceding_comment,
+            },
+        },
+        ValueDef::ModuleImport(module_import) => {
+            // TODO!
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/def.rs:369]: TODO

File: crates/compiler/fmt/src/def.rs [hunk 5]
```diff
@@ -51,6 +69,440 @@ impl<'a> Formattable for Defs<'a> {
...
+            SpacesBefore {
+                before: &[],
+                item: ValueDef::ModuleImport(module_import),
+            }
+        }
+        ValueDef::IngestedFileImport(ingested_file_import) => {
+            // TODO!
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/def.rs:376]: TODO

File: crates/compiler/fmt/src/def.rs [hunk 6]
```diff
@@ -51,6 +69,440 @@ impl<'a> Formattable for Defs<'a> {
...
+            SpacesAfter {
+                item: ValueDef::Expect {
+                    condition: arena.alloc(Loc::at(condition.region, condition_lifted.item)),
+                    preceding_comment,
+                },
+                after: condition_lifted.after,
+            }
+        }
+        ValueDef::ModuleImport(module_import) => {
+            // TODO!
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/def.rs:479]: TODO

File: crates/compiler/fmt/src/def.rs [hunk 7]
```diff
@@ -51,6 +69,440 @@ impl<'a> Formattable for Defs<'a> {
...
+            SpacesAfter {
+                item: ValueDef::ModuleImport(module_import),
+                after: &[],
+            }
+        }
+        ValueDef::IngestedFileImport(ingested_file_import) => {
+            // TODO!
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/def.rs:486]: TODO

File: crates/compiler/fmt/src/expr.rs [hunk 8]
```diff
@@ -199,129 +73,13 @@ impl<'a> Formattable for Expr<'a> {
...
-                // TODO: enforce in the type system that spaces must be pushed to the "outside".
-                // In other words, Expr::Apply should look something like the following, and there shouldn't be Expr::SpaceBefore and ::SpaceAfter.
-                //
-                // ```
-                // Apply(&'a SpaceAfter<Loc<Expr<'a>>>, &'a [&'a SpaceBefore<Loc<Expr<'a>>>], CalledVia),
-                // ```
-                //
-                // assert!(loc_expr.extract_spaces().before.is_empty(), "{:#?}", self);
+                let apply_needs_parens = parens == Parens::InApply;
+                // || (parens == Parens::InOperator && apply_has_closure(self));
```
Comment (lukewilliamboswell) [inline crates/compiler/fmt/src/expr.rs:77]: is this meant to be commented out?

File: crates/compiler/parse/src/normalize.rs [hunk 9]
```diff
@@ -714,32 +706,39 @@ impl<'a> Normalize<'a> for Expr<'a> {
...
-                defs.spaces.clear();
-
-                for type_def in defs.type_defs.iter_mut() {
-                    *type_def = type_def.normalize(arena);
-                }
+                // let mut defs = a.clone();
+                // defs.space_before = vec![Default::default(); defs.len()];
+                // defs.space_after = vec![Default::default(); defs.len()];
+                // defs.regions = vec![Region::zero(); defs.len()];
+                // defs.spaces.clear();
```
Comment (lukewilliamboswell) [inline crates/compiler/parse/src/normalize.rs:713]: Does this need to be removed?

File: crates/compiler/test_syntax/src/test_helpers.rs [hunk 10]
```diff
@@ -238,7 +238,9 @@ impl<'a> Input<'a> {
                 * * * AST after formatting:\n{:#?}\n\n",
                 self.as_str(),
                 output.as_ref().as_str(),
-                actual,
+                // actual,
```
Comment (lukewilliamboswell) [inline crates/compiler/test_syntax/src/test_helpers.rs:241]: does this need to be removed?

File: crates/compiler/test_syntax/tests/test_snapshots.rs [hunk 11]
```diff
@@ -631,7 +734,7 @@ mod test_snapshots {
             Err(err) => Err(format!("{err:?}")),
         };
 
-        println!("{:?}", result);
+        // println!("{:?}", result);
```
Comment (lukewilliamboswell) [inline crates/compiler/test_syntax/tests/test_snapshots.rs:737]: why was this removed?

Comment (joshuawarner32) [reply crates/compiler/test_syntax/tests/test_snapshots.rs:737]: I actually forget why I had commented this out, but pretty sure it was just for cleanliness when doing debugging.

File: crates/compiler/fmt/src/expr.rs (hunk 8 above)
Comment (joshuawarner32) [reply crates/compiler/fmt/src/expr.rs:77]: (removed)

**Review by lukewilliamboswell — APPROVED** [review]
